Simplify the expandability check

akirak opened the pull request:
Hello, thank you for creating this nice package. I am currently exploring its features. 

I noticed that some of the code can be simplified. Please have a look at this PR and consider it. Below is a summary:

- You can use `cl-ecase` instead of the combination of `cond` and `eq`.
- `after-symbol` doesn't seem to work as I expected. I think this should be `(looking-at (rx symbol-end))`.
- `looking-back` is slow, so it should be avoided wherever possible.

akirak committed:
Use cl-case to simplify the conditionals

akirak committed:
Simplify and optimize the test of fancy-dabbrev-expansion-context

(thing-at-point 'symbol) can return non-nil even if the position
is at the beginning or in the middle of a symbol. Use (rx symbol-end).

looking-back is slow, so it would be better to avoid it wherever possible.

jrosdahl commented:
Thanks for the pull request.

> I noticed that some of the code can be simplified. Please have a look at this PR and consider it. Below is a summary:
> 
> * You can use `cl-ecase` instead of the combination of `cond` and `eq`.

Reading up on `cl-ecase`, I notice that it returns an error if no branch matches, which is a change in behavior compared to the original `cond`s. Is this intentional? I think that I deliberately wanted to not give an error if the user has set an unsupported value for `fancy-dabbrev-expansion-context` or `fancy-dabbrev-preview-context`, especially since the code is run asynchronously with a timer.

> * `after-symbol` doesn't seem to work as I expected. I think this should be `(looking-at (rx symbol-end))`.

It's intentional that `after-symbol` matches in the middle of a symbol. Consider the case where you have `|bar` (with `|` being the cursor) and then type `foo` (giving `foo|bar`). If you then call `dabbrev-expand` it will expand the `foo` part (since `foo` is a symbol), so `fancy-dabbrev-expand` should do the same.

> * `looking-back` is slow, so it should be avoided wherever possible.

OK. Is it slow in practice even with the `(line-beginning-position)` limit? I can't find a slow case when testing this quickly.

akirak committed:
Revert "Simplify and optimize the test of fancy-dabbrev-expansion-context"

This reverts commit 8bcdbd31886835cc5b5b79e9bef265cc69cfe56d.

akirak commented:
>  I think that I deliberately wanted to not give an error if the user has set an unsupported value

I got your point. Now it uses `cl-case`, which returns nil when there is no matching value. Please take it freely if you find it syntactically better.

> It's intentional that after-symbol matches in the middle of a symbol. 

Sorry about this. I changed it because I expected something similar to completion in zsh (zsh doesn't fire completion in the middle of a word in my config). I reverted it. The option doesn't matter to me now (because I am using https://github.com/jrosdahl/fancy-dabbrev/issues/14#issuecomment-1055419199).

Regarding the performance of `looking-back`, I still think it is possible to avoid it, but it may be better to leave it as it is until someone finds a noticeable delay. I don't mind it now.

Thank you for your response.

akirak commented:
I'll close this, as the change is too trivial and does not affect the functionality.
